--- abcde-old-2.2.3	2005-03-08 12:44:01.000000000 -0500
+++ abcde2	2005-03-08 12:11:07.000000000 -0500
@@ -366,6 +366,9 @@
 	mpc)
 		run_command tagtrack-$1 true
 		;;
+	m4a)
+		run_command tagtrack-$1 true
+		;;
 	esac
 	done
 }
@@ -489,6 +492,10 @@
 				## FIXME ## to the encoder ends up empty.
 				run_command encodetrack-$OUTPUT-$1 nice $ENCNICE $MPPENCODER $MPPENCODEROPTS --artist "$TRACKARTIST" --album "$DALBUM" --title "$TRACKNAME" --track "$1" --genre "$CDGENRE" --year "$CDYEAR" --comment "$COMMENT" "$IN" "$OUT"
 				;;
+			m4a)	
+				# AAC/MPEG-4 format (.m4a) is done locally, with inline tagging.
+				run_command encodetrack-$OUTPUT-$1 nice $ENCNICE $M4AENCODER $M4AENCODEROPTS --artist "$TRACKARTIST" --album "$DALBUM" --title "$TRACKNAME" --track "$1" --genre "$CDGENRE" --year "$CDYEAR" --comment "$COMMENT" -w "$IN" -o "$OUT"
+				;;
 			esac
 		done
 		# Only remove .wav if the encoding succeeded
@@ -1695,6 +1702,7 @@
 SPEEXENCODERSYNTAX=default
 MPPENCODERSYNTAX=default
 NORMALIZERSYNTAX=default
+M4AENCODERSYNTAX=default
 
 OUTPUTFORMAT='${ARTISTFILE}-${ALBUMFILE}/${TRACKNUM}.${TRACKFILE}'
 # Use the following VAOUTPUTFORMAT to revert to 2.0.x VA format:
@@ -1741,6 +1749,7 @@
 SPEEXENC=speexenc
 # mpp (Musepack)
 MPPENC=mppenc
+M4AENC=faac
 
 ID3=id3
 ID3V2=id3v2
@@ -1776,6 +1785,7 @@
 SPEEXENCOPTS=
 # mpc
 MPPENCOPTS=
+M4AENCOPTS=
 
 ID3OPTS=
 ID3V2OPTS=
@@ -2042,6 +2052,7 @@
 		        ;;
 		spx) [ "$SPEEXENCODERSYNTAX" = "default" ] && SPEEXENCODERSYNTAX=speexenc ;;
 		mpc) [ "$MPPENCODERSYNTAX" = "default" ] && MPPENCODERSYNTAX=mppenc ;;
+		m4a) [ "$M4AENCODERSYNTAX" = "default" ] && M4AENCODERSYNTAX=faac ;;
 		*) echo "abcde error: Invalid OUTPUTTYPE defined" >&2
 		   exit 1
 		   ;;
@@ -2103,6 +2114,12 @@
 		MPPENCODER="$MPPENC"
 		;;
 esac
+case "$M4AENCODERSYNTAX" in
+	faac)
+		M4AENCODEROPTS="$M4AENCOPTS"
+		M4AENCODER="$M4AENC"
+		;;
+esac
 
 # and which tagger
 
@@ -2147,7 +2164,7 @@
 	${NEEDCOMMENTER+$VORBISCOMMENT} ${NEEDMETAFLAC+$METAFLAC} \
 	${NEEDNORMALIZER+$NORMALIZER} ${NEEDEJECT+$EJECT} \
 	${NEEDDISKTOOL+disktool} ${NEEDCDSPEED+$CDSPEED} \
-	${NEEDVORBISGAIN+$VORBISGAIN}
+	${NEEDVORBISGAIN+$VORBISGAIN} $M4AENCODER 
 do
 	# Cut off the command-line options we just added in
 	X=$(echo $X | cut -d' ' -f2)
