Index: abcde
===================================================================
--- abcde	(revision 209)
+++ abcde	(working copy)
@@ -452,10 +452,18 @@
 				ARTISTFILE="$(mungefilename "$TRACKARTIST")"
 				ALBUMFILE="$(mungefilename "$DALBUM")"
 				gettracknum
-				if [ "$VARIOUSARTISTS" = "y" ]; then
-					OUTPUTFILE="$(eval echo "$VAOUTPUTFORMAT")"
-				else
-					OUTPUTFILE="$(eval echo "$OUTPUTFORMAT")"
+				if [ "$ONETRACK" = "y" ]; then 
+					if [ "$VARIOUSARTISTS" = "y" ]; then
+						OUTPUTFILE="$(eval echo \""$VAONETRACKOUTPUTFORMAT"\")"
+					else
+						OUTPUTFILE="$(eval echo \""$ONETRACKOUTPUTFORMAT"\")"
+					fi
+				else	
+					if [ "$VARIOUSARTISTS" = "y" ]; then
+						OUTPUTFILE="$(eval echo \""$VAOUTPUTFORMAT"\")"
+					else
+						OUTPUTFILE="$(eval echo \""$OUTPUTFORMAT"\")"
+					fi
 				fi
 				OUTPUTFILES[$REPLAYINDEX]="$OUTPUTDIR/$OUTPUTFILE.$OUTPUT"
 				(( REPLAYINDEX = $REPLAYINDEX + 1 ))
@@ -1240,10 +1248,18 @@
 		gettracknum
 		# Supported variables for OUTPUTFORMAT are GENRE, YEAR, ALBUMFILE,
 		# ARTISTFILE, TRACKFILE, and TRACKNUM.
-		if [ "$VARIOUSARTISTS" = "y" ]; then
-			OUTPUTFILE="$(eval echo "$VAOUTPUTFORMAT")"
-		else
-			OUTPUTFILE="$(eval echo "$OUTPUTFORMAT")"
+		if [ "$ONETRACK" = "y" ]; then 
+			if [ "$VARIOUSARTISTS" = "y" ]; then
+				OUTPUTFILE="$(eval echo \""$VAONETRACKOUTPUTFORMAT"\")"
+			else
+				OUTPUTFILE="$(eval echo \""$ONETRACKOUTPUTFORMAT"\")"
+		    	fi
+		else	
+			if [ "$VARIOUSARTISTS" = "y" ]; then
+				OUTPUTFILE="$(eval echo \""$VAOUTPUTFORMAT"\")"
+			else
+				OUTPUTFILE="$(eval echo \""$OUTPUTFORMAT"\")"
+			fi
 		fi
 		if checkerrors "tagtrack-$OUTPUT-$1"; then :; else
 			# Once we know the specific output was successful, we can change
